lib/kafka/server.go

```diff
@@ -2,85 +2,121 @@ package kafka
+		"kafka-default", []string{config.Address}, kafka.NewBrokerConf("discord-punt"))
 	if err != nil {
-		panic(err)
+		log.Panicf("Failed to create Kafka broker: %s", err)
 	}
 
-	server := Server{
+	return &Server{
 		Messages: messages,
-		Errors: errors,
-		Topic: config.Topic,
-		Broker: broker,
+		Errors:   errors,
+		Topic:    config.Topic,
+		Broker:   broker,
 	}
-	return &server
 }
 
+// Start begins reading from Kafka and processing messages.
 func (s *Server) Start() {
 	partitions, err := s.Broker.PartitionCount(s.Topic)
 	if err != nil {
-		log.Panic("Could not get broker partition count: %v", err);
+		log.Panicf("Could not get broker partition count: %v", err)
 	}
 
 	for partitionID := int32(0); partitionID < partitions; partitionID++ {
 		go s.partitionReader(s.Topic, partitionID)
 	}
 }
 
-func (s* Server) partitionReader(topic string, partitionID int32) {
-	conf := kafka.NewConsumerConf(topic, partitionID)
-	consumer, err := s.Broker.Consumer(conf)
-	if err != nil {
-		s.Errors <- syslog.InvalidMessage{Data: "Could not get Kafka consumer", Error: err}
+func (s *Server) partitionReader(topic string, partitionID int32) {
+	connectRetry := &backoff.Backoff{
+		Min:    time.Second,
+		Max:    time.Minute,
+		Jitter: true,
+	}
+
+	messageRetry := &backoff.Backoff{
+		Min:    time.Millisecond,
+		Max:    time.Second,
+		Jitter: true,
 	}
+
 	for {
-		msg, err := consumer.Consume()
+		conf := kafka.NewConsumerConf(topic, partitionID)
+		consumer, err := s.Broker.Consumer(conf)
 		if err != nil {
-			s.Errors <- syslog.InvalidMessage{Data: "Failed to condume message", Error: err}
+			//s.Errors <- syslog.InvalidMessage{Data: "Failed to condume message", Error: err}
+			log.Printf("Failed to connect to Kafka %s:%d: %s", topic, partitionID, err)
+			time.Sleep(connectRetry.Duration())
+			continue
 		}
+		connectRetry.Reset()
 
-		unpack := make(map[string]interface{})
-		if err := json.Unmarshal(msg.Value, &unpack); err != nil {
-			s.Errors <- syslog.InvalidMessage{Data: "Invalid message received", Error: err}
-		}
-		var timestamp Timestamp
-		if err := json.Unmarshal([]byte(unpack["data"].(string)), &timestamp); err != nil {
-			s.Errors <- syslog.InvalidMessage{Data: "Could not get message timestamp", Error: err}
-		}
+		for {
+			msg, err := consumer.Consume()
+			if err != nil {
+				log.Printf("Failed to consume event from partition %d: %s", partitionID, err)
+				time.Sleep(messageRetry.Duration())
+				continue
+			}
+
+			// All messages should be Freight message formatted messages, unwrap that first
+			// and by default use the SendTime of the message as the timestamp.
+			var message freight.Message
+			if err := json.Unmarshal(msg.Value, &message); err != nil {
+				log.Printf("Failed to unmarshal wrapper: %s", err)
+				time.Sleep(messageRetry.Duration())
+				continue
+			}
+			timestamp := message.SendTime
+
+			// Some messages contain a client submitted "timestamp" field. If that's provided,
```

Comment:
All messages should have client timestamp.

Reply:
The `rtc_quality_report` data does right now but punt also works for internal producers that need to write data to Kafka and I want punt to be able to consume it even if it's not a client generated message.

Reply:
This is probably fine for now, I think in the future we'll probably want to move this elsewhere in punt so that you can configure given payloads to provide their own timestamps.